design/hostclaim-multitenancy-and-hybrid-clusters.md

```diff
@@ -0,0 +1,681 @@
+standard baremetal cluster definitions is the presence of a ``kind`` field in
+the Metal3Machine templates. The value of this field is set to ``baremetal``.
+
+When the cluster is started, a HostClaim is created for each Metal3Machine
+associated to the cluster. The ``hostSelector`` and ``kind`` fields are
+inherited from the Metal3Machine. They are used to define the BareMetalHost
+associated with the cluster. The associated BareMetalHost is not in the same
+namespace as the HostClaim. The exact definition of the BareMetalHost remains
+hidden from the cluster user, but parts of its status and metadata are copied
+back to the Host namespace. With this information,
+the data template controller has enough details to compute the different
+secrets (userData, metaData and networkData) associated to the Metal3Machine.
+Those secrets are linked to the HostClaim and, ultimately, to the
+BareMetalHost.
+
+When the cluster is modified, new Machine and Metal3Machine resources replace
+the previous ones. The HostClaims follow the life-cycle of the Metal3Machines
+and are destroyed unless they are tagged for node reuse. The BareMetalHosts are
+recycled and are bound to new HostClaims, potentially belonging to other
+clusters.
+
+#### Hybrid Clusters
+
+As a cluster administrator I would like to build a cluster with different kind of nodes.
+
+This scenario assumes that:
+
+* the cloud technologies CT_i use qcow images and cloud-init to
+  define workloads.
+* Clouds C_i implementing CT_i are accessible through
+  credentials and endpoints described in a resource Cr_i.
+* a HostClaim controller exists for each CT_i. Compute resource can
+  be parameterize through arguments in the HostClaim arg_ij.
+
+One can build a cluster where each machine deployment MD_i
+contains Metal3Machine templates referring to kind CT_i.
+The arguments identify the credentials to use (CR_i)
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: md-i
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: dt-i
+      kind: CT_i
+      args:
+        arg_i1: v1
+        ...
+        arg_ik: vk
+      hostSelector:
+        matchLabels:
+          ...
+      image:
+        checksum: https://image_url.qcow2.md5
+        format: qcow2
+        url: https://image_url.qcow2.md5
+```
+
+The Metal3Machine controllers will create HostClaims with different kinds
+handled by different controllers creating the different compute resources.
+Connectivity must be established between the different subnets where each
+controller creates its compute resources.
+
+The argument extension is not strictly necessary but is cleaner than using
+matchLabels in specific domains to convey information to controllers.
+Controllers for disposable resources such as virtual machine typically do not
+use hostSelectors. Controllers for a "bare-metal as a service" service
+may use selectors.
+
+#### Compute Resources in Another Cluster
```

Comment:
This looks like a feature that could be implemented separately from the larger HostClaim discussion.

Reply:
I think it will be #429 and be included in capm3 code. For HostClaim you need to add the resource to the list of tracked remote resources (BMH in the proposal). I suggested to use another hostclaim controller... but this was accepting that the inherent modularity of hostclaims which is key for hybrid clusters was a good idea, so it was rejected. An example implementation along #429 principle is [here](https://gitlab.com/Orange-OpenSource/kanod/cluster-api-provider-metal3/-/commit/70088f43686792c25603013ebd47e41e399a8eca) (for hostclaims only, not bmh).